### PR TITLE
DEV: Add codespell to pre-commit hooks

### LINKED ISSUE #22740: [MNT]: Add codespell to pre-commit hooks
### Summary

In PR #22733 @QuLogic caught a spelling error of mine (which is great!) but there will inevitably be typos that slip through if only humans are checking. https://github.com/codespell-project/codespell is a really nice way to check for typos and it has a nice pre-commit hook. As Matplotlib [already uses pre-commit](https://github.com/matplotlib/matplotlib/blob/da9533d507c1f06ceee5078c9cbbd99897809d31/.pre-commit-config.yaml) :rocket: (thanks to @ianhi in PR #21583) this would be an easy add. :+1: 

### Proposed fix

Add the following hook (maybe some variations) to [`.pre-commit-config.yaml`](https://github.com/matplotlib/matplotlib/blob/da9533d507c1f06ceee5078c9cbbd99897809d31/.pre-commit-config.yaml)

```yaml
  - repo: https://github.com/codespell-project/codespell
    rev: v2.1.0
    hooks:
      - id: codespell
        files: ^.*\.(py|md|rst|yml)$
```

### PR DESCRIPTION
**Pull request by matthewfeickert**
## PR Summary

Resolves #22740

Add https://github.com/codespell-project/codespell as a pre-commit hook that checks for common misspellings of English words in files with the following extensions: `.py`, `.c`, `.cpp`, `.h`, `.m`, `.md`, `.rst`, `.yml`.

Ignore instances of:
* 'ans', 'axises', 'curvelinear', 'hist', 'nd', 'oly', 'thisy', 'wit'
* 'ba' for bound axis and bound args
* 'cannotation' for centered annotation
* 'coo' for COO (Coordinate list)
* 'flate' for Flate compression
* 'inout' for lib/matplotlib/rcsetup.py
* 'ment' for alignment with formatting
* 'whis' for whikser plot
* ~'ot' for offset transform~ (Resolved in PR #22784)
* 'sur' for Big Sur
* 'TE' for triangle elements

and skip checking [doc/users/project/credits.rst](https://github.com/matplotlib/matplotlib/blob/8b1881fd49b49bf85a7b91575f4653be41c26294/doc/users/project/credits.rst) as the list of names of contributors isn't worth adding in to the ignore file to avoid false positives.

The above is what is needed to get 

```console
$ pre-commit run codespell --all-files
codespell................................................................Passed
```

on the current `HEAD` of `main`.

Here only the `codespell` options `--ignore-words` and `--skip` are used, and the usual addition of `--write-changes` is ignored, as it seems from PR #21583 that automatic corrections were not desired.

## PR Checklist

**Tests and Styling**
- [n/a] Has pytest style unit tests (and `pytest` passes).
- [n/a] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [n/a] New features are documented, with examples if plot related.
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (matthewfeickert) [inline .pre-commit-config.yaml:37]: The list hopefully doesn't seem too excessive, but as Matplotlib has a lot of abbreviations of variables in it this is the minimum number of ignores needed to get `codespell` working when run across all of the above file types.

**Comment by matthewfeickert** [comment]
> I like the idea! Will this enforce not having any errors? So if one introduce a new abbreviation, it must also go in the ignore list? (I guess it is rare, but just want to understand.)

@oscargus No, it is only for character sets that `codespell` thinks is a mispelling of a common word. For example, if you had a variable named `algo` things would be fine

```diff
diff --git a/lib/matplotlib/artist.py b/lib/matplotlib/artist.py
index 1f33b9d3ec..a301a257f0 100644
--- a/lib/matplotlib/artist.py
+++ b/lib/matplotlib/artist.py
@@ -59,6 +59,7 @@ def allow_rasterization(draw):
                 renderer.stop_rasterizing()
                 renderer.start_rasterizing()
 
+    algo = 1
     draw_wrapper._supports_rasterization = True
     return draw_wrapper

```
```console
$ pre-commit run codespell --all-files
codespell................................................................Passed
```

but if you had a variable named `mapp` it would think you had typoed "map"

```diff
diff --git a/lib/matplotlib/artist.py b/lib/matplotlib/artist.py
index 1f33b9d3ec..a543035392 100644
--- a/lib/matplotlib/artist.py
+++ b/lib/matplotlib/artist.py
@@ -59,6 +59,7 @@ def allow_rasterization(draw):
                 renderer.stop_rasterizing()
                 renderer.start_rasterizing()
 
+    mapp = 1
     draw_wrapper._supports_rasterization = True
     return draw_wrapper
```
```console
$ pre-commit run codespell --all-files
codespell................................................................Failed
- hook id: codespell
- exit code: 65

lib/matplotlib/artist.py:62: mapp ==> map

```

and then you would need to add `mapp` to the ignore file if you really wanted this variable name.

**Comment by QuLogic** [comment]
> * 'coo' for COO (Coordinate list) 
> * 'sur' for Big Sur

Would it fix these by capitalizing them correctly instead? They appear to be comments or docstrings that can easily be fixed.

> * 'flate' for Flate compression

I can't find this.

You should also add (if not already automatically ignored by e.g., `.gitignore`), `extern`, `build`, and `lib/matplotlib/backends/web_backend/node_modules`.

**Comment by QuLogic** [comment]
> 'ot' for offset transform

All of these can be written out as a full word, so I've opened #22784 doing so.

**Comment by matthewfeickert** [comment]
> > * 'coo' for COO (Coordinate list)
> > * 'sur' for Big Sur
> 
> Would it fix these by capitalizing them correctly instead? They appear to be comments or docstrings that can easily be fixed.

@QuLogic Sadly no, as `--ignore-words-list` is a

> comma separated list of words to be ignored by codespell. Words are case sensitive based on how they are written in the dictionary file.

and from https://github.com/codespell-project/codespell/issues/2026 it doesn't really seem like this is possible to have case sensitive ignores. :(

`codespell` dictionaries are slightly confusingly named as they are lookup of spelling _corrections_, and not words, passed in with `--dictionary` which is a

> custom dictionary file that contains spelling corrections. If this flag is not specified or equals "-" then the default dictionary is used. This option can be specified multiple times.

So as these aren't typos, but just abbreviations or words that `codespell` thinks _might_ be typos, there isn't really a good alternative at the moment except to ignore them (as far as I understand).

> > * 'flate' for Flate compression
> 
> I can't find this.

https://github.com/matplotlib/matplotlib/blob/9905cf024035f4aaecaebeab7ad2d9fa35431fe4/lib/matplotlib/backends/backend_pdf.py#L1651

> You should also add (if not already automatically ignored by e.g., `.gitignore`), `extern`, `build`, and `lib/matplotlib/backends/web_backend/node_modules` to be ignored.

Correct, as this is a pre-commit hook then it only runs over files that are under version control so these don't need to be worried about as you already have them ignored. :+1:

Example:

```console
$ echo "mapp" > example.txt
$ pre-commit run codespell --all-files
codespell................................................................Passed
```
 
> All of these can be written out as a full word, so I've opened #22784 doing so.

Awesome and thank you! I didn't want to presume that things should be modified, so I didn't open make any changes in this PR. Once PR #22784 is merged I'll rebase this on one it. :+1:

**Comment by QuLogic** [comment]
> > * 'flate' for Flate compression
> 
> I can't find this.

Oh, because it's case-insensitive, okay.

> > You should also add (if not already automatically ignored by e.g., `.gitignore`), `extern`, `build`, and `lib/matplotlib/backends/web_backend/node_modules` to be ignored.
> 
> Correct, as this is a pre-commit hook then it only runs over files that are under version control so these don't need to be worried about as you already have them ignored. +1

The `extern` directory _is_ in version control.

**Comment by matthewfeickert** [comment]
> The `extern` directory _is_ in version control.

Ah yeah. whoops! Okay fixed that locally on my fork and I'll push that in the rebase after PR #22784 is in.

Another note: Given https://github.com/codespell-project/codespell/issues/2063 I can't add comments to the ignore file to make it clear why these words are ignored. So the commit message and this PR will have to do for the time being

**Comment by jklymak** [comment]
I'm always confused by these pre-commit hooks as I don't currently use them.  Won't our CI will have to run this as well, otherwise folks who don't use pre-commit hooks will fail folks who do?  

But overall, I'm not clear what problem adding this is supposed to fix.  No doubt just general ignorance on my part.

**Comment by matthewfeickert** [comment]
> I'm always confused by these pre-commit hooks as I don't currently use them. Won't our CI will have to run this as well, otherwise folks who don't use pre-commit hooks will fail folks who do?

Hi @jklymak. :wave: There was back and forth discussion on PR #21583 about enabling [pre-commit.ci](https://pre-commit.ci/), and while @ianhi enabled CI running in the config but disabled pushing back fixes (c.f. https://github.com/matplotlib/matplotlib/pull/21583#issuecomment-973295547)

https://github.com/matplotlib/matplotlib/blob/a436253d1d672cd46b0da93acdd868d792f2dbb8/.pre-commit-config.yaml#L1-L2

it apparently never got turned on as

[pre-commit.ci status](https://results.pre-commit.ci/latest/github/matplotlib/matplotlib/main)

is unknown at the moment.


My take is that pre-commit.ci is well worth the "[yet another service](https://github.com/matplotlib/matplotlib/pull/21583#issuecomment-973402576)" Ryan was unsure about, as it makes it viable for me to actually maintain pre-commit setups across many GitHub organizations, and to make sure that things are actually getting run. As a pre-commit.ci heavy user I'm very happy to discuss this if you have questions.

So yes, to get the full benefit you should run it in CI, but this is easy to do.

> But overall, I'm not clear what problem adding this is supposed to fix. No doubt just general ignorance on my part.

pre-commit in general? Or the addition of `codespell`? If the latter, then automatically catching and notifying you at commit time that you probably have a typo in the English language parts of the code and docs. So typo avoidance for free and one less thing that reviewers need to do. c.f. https://github.com/matplotlib/matplotlib/pull/22733#discussion_r839101131 as an example.

**Comment by jklymak** [comment]
No I'm fine with pre-commit as an idea.  I just am unsure about codespell.  Certainly if it works on our docs and docstrings that is helpful.  I was less certain about its use in code per-se. 

However it seems we need to get pre-commit to run properly on our CI first?

**Comment by matthewfeickert** [comment]
> However it seems we need to get pre-commit to run properly on our CI first?

This just requires a maintainer turning it (https://pre-commit.ci/) on. There's nothing that needs to be done on the GitHub side.

**Comment by jklymak** [comment]
Great - I assume it currently passes?   I'll ask on gitter if we can turn it on.

**Comment by matthewfeickert** [comment]
> Great - I assume it currently passes?

This PR passes as

```console
$ pre-commit run codespell --all-files
codespell................................................................Passed
```

but I believe (I could be wrong, as I'm used to having `pre-commit.ci` run with `autofix_prs: true`) that `pre-commit.ci` runs `pre-commit run --all-files`, which currently fails

```console
$ pre-commit run --all-files
Check for added large files..............................................Passed
Check docstring is first.................................................Failed
- hook id: check-docstring-first
- exit code: 1

lib/matplotlib/dates.py:225 Multiple module docstrings (first docstring on line 1).
lib/matplotlib/_constrained_layout.py:28 Multiple module docstrings (first docstring on line 1).
lib/matplotlib/backend_tools.py:992 Multiple module docstrings (first docstring on line 1).

Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing plot_types/README.rst
Fixing doc/api/transformations.rst
Fixing lib/matplotlib/tests/README
Fixing doc/devel/testing.rst
Fixing doc/users/next_whats_new/modify_stairs_fill_edge_behaviour.rst
Fixing doc/devel/style_guide.rst
Fixing doc/users/next_whats_new/windows_arm64.rst
Fixing plot_types/basic/README.rst
Fixing doc/api/next_api_changes/deprecations/22268-OG.rst
Fixing doc/api/toolkits/mplot3d/faq.rst
Fixing .github/workflows/tests.yml
Fixing doc/_static/switcher.json
Fixing doc/_static/.gitignore
Fixing doc/api/next_api_changes/deprecations/22345-JK.rst
Fixing LICENSE/LICENSE_COURIERTEN
Fixing doc/users/next_whats_new/inset_axes_improvements.rst
Fixing plot_types/arrays/README.rst
Fixing LICENSE/LICENSE_BAKOMA
Fixing doc/api/lines_api.rst
Fixing LICENSE/LICENSE
Fixing doc/api/next_api_changes/development/22205-ES.rst
Fixing doc/api/next_api_changes/deprecations/22134-OG.rst
Fixing doc/api/_enums_api.rst
Fixing lib/matplotlib/backends/web_backend/css/page.css
Fixing plot_types/unstructured/README.rst
Fixing doc/api/next_api_changes/removals/21980-CC.rst
Fixing doc/users/next_whats_new/list_font_names.rst
Fixing doc/api/next_api_changes/behavior/22691-JMK.rst
Fixing LICENSE/LICENSE_CARLOGO
Fixing plot_types/stats/README.rst
Fixing doc/api/next_api_changes/removals/22516-OG.rst
Fixing doc/api/next_api_changes/behavior/22639-RA.rst
Fixing README.rst
Fixing doc/users/next_whats_new/layout_engine.rst
Fixing doc/api/next_api_changes/deprecations/22133-OG.rst
Fixing doc/missing-references.json
Fixing LICENSE/LICENSE_STIX

Mixed line ending........................................................Passed
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing doc/users/next_whats_new/3d_plot_roll_angle.rst
Fixing doc/devel/contributing.rst
Fixing LICENSE/LICENSE_AMSFONTS
Fixing doc/users/next_whats_new/modify_stairs_fill_edge_behaviour.rst
Fixing .github/ISSUE_TEMPLATE/documentation.yml
Fixing doc/_templates/automodule.rst
Fixing doc/README.txt
Fixing doc/users/index.rst
Fixing .github/ISSUE_TEMPLATE/maintenance.yml
Fixing doc/users/explain/fonts.rst
Fixing doc/api/tri_api.rst
Fixing CODE_OF_CONDUCT.md
Fixing doc/api/next_api_changes/development/00001-ABC.rst
Fixing LICENSE/LICENSE_BAKOMA
Fixing doc/api/lines_api.rst
Fixing doc/users/github_stats.rst
Fixing doc/users/faq/index.rst
Fixing doc/api/next_api_changes/removals/00001-ABC.rst
Fixing LICENSE/LICENSE_COLORBREWER
Fixing doc/api/next_api_changes/behavior/00001-ABC.rst
Fixing src/_backend_agg.h
Fixing plot_types/unstructured/README.rst
Fixing doc/users/resources/index.rst
Fixing doc/api/next_api_changes/removals/21980-CC.rst
Fixing .github/ISSUE_TEMPLATE/bug_report.yml
Fixing lib/matplotlib/backends/web_backend/css/fbm.css
Fixing examples/event_handling/README.txt
Fixing doc/users/next_whats_new/extending_MarkerStyle.rst
Fixing doc/api/patches_api.rst
Fixing LICENSE/LICENSE_STIX

flake8...................................................................Passed
codespell................................................................Passed
```

which maybe explains @tacaswell's comment here https://github.com/matplotlib/matplotlib/pull/21583#issuecomment-1022770988 more.

As these failing hooks are across a large number of files, if it would make it easier in a follow up PR to this one (or a new PR in which this is done in advance of this getting merged) I could apply them iand then add the commit to a `.git-blame-ignore-revs` which would then [ignore those commits in the blame view](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) (c.f. https://github.com/scikit-build/scikit-build/pull/665#issuecomment-1076806751 for how `scikit-build` uses this to [avoid having large applications of `black` show up in blame](https://github.com/scikit-build/scikit-build/blob/d90c4cc0949beb055dc78ac5d5dfc222be1c112d/.git-blame-ignore-revs)). Or if that's not acceptable the

https://github.com/matplotlib/matplotlib/blob/a436253d1d672cd46b0da93acdd868d792f2dbb8/.pre-commit-config.yaml#L22

could get removed (https://github.com/matplotlib/matplotlib/pull/21583#issuecomment-1022646596).

**Review by jklymak — CHANGES_REQUESTED** [review]
Just blocking until we sort out how top make sure the pre-commit doesn't get out of sync with CI.  So far options seem to be run precommit-ci or make reviewdog run a parallel set of checks.  

The point is that we shouldn't merge PRs that fail the precommit-ci or we will have trivial follow-up PRs to correct things for pre-commit folks.

**Comment by matthewfeickert** [comment]
> Just blocking until we sort out how top make sure the pre-commit doesn't get out of sync with CI. So far options seem to be run precommit-ci or make reviewdog run a parallel set of checks.
> 
> The point is that we shouldn't merge PRs that fail the precommit-ci or we will have trivial follow-up PRs to correct things for pre-commit folks.

:+1: Makes sense. Comments welcome on PR #22809 which should address all of this.

File: .pre-commit-config.yaml
Comment (greglucas) [inline .pre-commit-config.yaml:34]: Did you try `.cpp` and `*.m` for macos as well? https://github.com/matplotlib/matplotlib/blob/main/src/_macosx.m

It looks like the extern directory is ignored, so hopefully this would only catch the cpp/objective-c that someone on the project has added.

Comment (matthewfeickert) [reply .pre-commit-config.yaml]: I had just selected the files that got caught when I ran `codespell` over all files, but I've added `.cpp` and `.m` into the extensions to run over to be future safe in the push I just did with the rebase. Also, yes, `extern/` is ignored for all of pre-commit and so safe from getting hit by this hook.

**Review by greglucas — APPROVED** [review]

**Comment by matthewfeickert** [comment]
@jklymak @greglucas @oscargus Now that PR #22809 has been merged in I've rebased this PR on that and all the hooks are passing.

```console
$ git rebase upstream/main 
Current branch mnt/add-codespell-to-pre-commit-hooks is up to date.
$ pre-commit run --all-files
check for added large files..............................................Passed
check docstring is first.................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
flake8...................................................................Passed
codespell................................................................Passed
```

So this is ready for review again.

**Review by greglucas — APPROVED** [review]

**Comment by tacaswell** [comment]
I'm in favor of getting this running on CI in _any_ way that marks with a red-x, merging this, and then iterating on how best to notify at what needs to be changed.

**Comment by matthewfeickert** [comment]
> I'm in favor of getting this running on CI in _any_ way that marks with a red-x

@tacaswell If you want CI running with pre-commit hooks now and to sort out what solution or combinations you want later then all you have to do is turn on pre-commit.ci: https://pre-commit.ci/


> merging this, and then iterating on how best to notify at what needs to be changed.

SGTM. If you haven't already please see https://github.com/matthewfeickert/matplotlib/pull/12 for an example of how both pre-commit.ci and reviewdog look and the differences in their actions. I'll also note that there's nothing keeping both pre-commit.ci and reviewdog running at the same time. I am willing to help implement _whatever the dev team wants_, there has just been no consensus.

**Comment by tacaswell** [comment]
OK, I have turned pre-commit on for the main repo (it was already on for pytest-mpl and ipympl).  I am not super stoked that it has the ability to push to the main repo, but it is not the only application with similar access.

**Comment by matthewfeickert** [comment]
> OK, I have turned pre-commit on for the main repo (it was already on for pytest-mpl and ipympl).

And everything on `main` is passing:

[pre-commit.ci status](https://results.pre-commit.ci/latest/github/matplotlib/matplotlib/main)

and things are looking pretty good across the board

[pre-commit-matplotlib](https://results.pre-commit.ci/repo/github/1385122)


> I am not super stoked that it has the ability to push to the main repo, but it is not the only application with similar access.

@tacaswell I think you already know this, so forgive me if I am over explaining, but this is because it has the ability to push back fixes if desired. As discussed before, this is turned off in the `pre-commit` config:

https://github.com/matplotlib/matplotlib/blob/e9e62ea5f0214fc24bdc07db80b1b3a8e5f4653f/.pre-commit-config.yaml#L1-L2

and as shown in https://github.com/matthewfeickert/matplotlib/pull/9 it will only attempt to push back changes to a PR if it is explicitly asked for them.

You mentioned that it isn't the only application that has permissions to do this, but if the dev team isn't comfortable with this then you can simply disable it the same way you turned it on and a different approach can be used.

As I imagine there might be some future discussions on this I can open up a GitHub Issue for what is the desired CI system moving forward.

**Comment by tacaswell** [comment]
Overly permissive OAuth tokens on my mind due to the recent travis/heroku issue. 

We do not have any private repos (which is what those attackers were going for), but you could in principle do something "interesting" if you could get your hands on a write token to the Matplotlib repo (and by "interesting" I mean "push malicious code").  However, unless it has _admin_ write privileges all of the branches that are "special" (`main`, `vX.Y.x`, and `*-doc`) require a PR to push to.

**Comment by matthewfeickert** [comment]
> However, unless it has _admin_ write privileges all of the branches that are "special" (`main`, `vX.Y.x`, and `*-doc`) require a PR to push to.

Indeed, having admin privileges would be a security disaster.

I would agree that the GitHub permissions listing is not super explicit about which permissions level an associated app has

pre-commit-permissions

but I can tell you that pre-commit.ci does not have admin privileges on any of the orgs that I'm an owner of (e.g. Scikit-HEP).

**Comment by matthewfeickert** [comment]
Hm. PR #22964 caught additional spelling errors. A large chunk of them I'm not surprised that `codespell` didn't catch as `doc/api/prev_api_changes` should be excluded by

https://github.com/matplotlib/matplotlib/blob/f8cd2c9f532f65f8b2e3dec6d54e03c48721233c/.pre-commit-config.yaml#L10-L11

but I'm not really sure why `ploted` -> `plotted` wasn't caught in 

https://github.com/matplotlib/matplotlib/blob/f8cd2c9f532f65f8b2e3dec6d54e03c48721233c/examples/pyplots/pyplot_simple.py#L6

`codespell` does describe itself as trying to catch common misspellings

> Fix common misspellings in text files. It's designed primarily for checking misspelled words in source code, but it can be used with other files as well. It does not check for word membership in a complete dictionary, but instead looks for a set of common misspellings. Therefore it should catch errors like "adn", but it will not catch "adnasdfasdf". This also means it shouldn't generate false-positives when you use a niche term it doesn't know about.

so I would have thought it would get things like this.

**Comment by greglucas** [comment]
I still think this is a net-positive to add. It will catch many common mistakes early on without a dev needing to comment on nitpick wording/styling updates.

**Comment by matthewfeickert** [comment]
> It will catch many common mistakes early on without a dev needing to comment on nitpick wording/styling updates.

Agreed @greglucas. The `codespell` team merged my `ploted->plotted` PR very quickly, which makes me think that I could probably open up a PR to them that get most of the typos identified in PR #22964 so that in subsequent releases of `codespell` the pre-commit hook would also catch these.

Also the next release of `codespell` fixes the Big Sur issue (c.f. https://github.com/codespell-project/codespell/issues/2299) so that could get removed from `ci/codespell-ignore-words.txt`.

**Comment by matthewfeickert** [comment]
@jklymak gentle ping on this if you have time this week.

**Review by jklymak — APPROVED** [review]
Sorry, been teaching an intensive course.  Didn't mean to keep th block on if others are ok with this.  Thanks for your patience.

**Comment by matthewfeickert** [comment]
> Sorry, been teaching an intensive course. Didn't mean to keep th block on if others are ok with this. Thanks for your patience.

@jklymak Nothing to be sorry about! Thanks very much for your continued and thoughtful feedback during the life cycle of this PR. :+1:

**Pull request merged**
